src/py42/services/_connection.py

```diff
@@ -0,0 +1,258 @@
+        headers=None,
+        cookies=None,
+        files=None,
+        auth=None,
+        hooks=None,
+        stream=False,
+        timeout=60,
+        cert=None,
+        proxies=None,
+    ):
+        response = None
+        for _ in range(2):
+            request = self._prepare_request(
+                method,
+                url,
+                params=params,
+                data=data,
+                json=json,
+                headers=headers,
+                cookies=cookies,
+                files=files,
+                auth=auth,
+                hooks=hooks,
+            )
+
+            response = self._session.send(
+                request,
+                stream=stream,
+                timeout=timeout,
+                verify=settings.verify_ssl_certs,
+                cert=cert,
+                proxies=proxies,
+            )
+
+            if not stream and response is not None:
+                # setting this manually speeds up read times
+                response.encoding = u"utf-8"
+
+            if response is not None and 200 <= response.status_code <= 399:
+                return Py42Response(response)
+
+            if response is not None and response.status_code == 401:
+                if isinstance(self._auth, C42RenewableAuth):
+                    self._auth.clear_credentials()
+
+        # if nothing has been returned after two attempts, something went wrong
+        _handle_error(method, url, response)
+
+    def _prepare_request(
+        self,
+        method,
+        url,
+        params=None,
+        data=None,
+        json=None,
+        headers=None,
+        cookies=None,
+        files=None,
+        auth=None,
+        hooks=None,
+    ):
+        url = urljoin(self.host_address, url)
+        self._session.proxies = settings.proxies
+        self._session.verify = settings.verify_ssl_certs
+
+        if json is not None:
+            data = json_lib.dumps(json)
+
+        user_headers = {u"User-Agent": settings.get_user_agent_string()}
+        if headers:
+            user_headers.update(headers)
+        self._headers.update(user_headers)
+        request = Request(
+            method=method,
+            url=url,
+            headers=self._headers,
+            files=files,
+            data=data,
+            params=params,
+            auth=auth or self._auth,
+            cookies=cookies,
+            hooks=hooks,
+        )
+
+        _print_request(method, url, params=params, data=data)
+        return self._session.prepare_request(request)
+
+    def _get_host_address(self):
+        if not self._host_address:
+            with self._resolve_lock:
+                if not self._host_address:
+                    host = self._host_resolver.get_host_address()
+                    self._init_host_info(host)
+        return self._host_address
+
+    def _init_host_info(self, host):
+        if not host.startswith(u"http://") and not host.startswith(u"https://"):
```

Comment:
ha, is non-encrypted http really allowed?

Reply:
always has been to support on prem, where whether or not you do that is totally up to them. it of course will not work at all when targeting cloud.